converter/interceptor.go

```diff
@@ -0,0 +1,215 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package converter
+
+import (
+	"context"
+
+	"go.temporal.io/api/command/v1"
+	commonpb "go.temporal.io/api/common/v1"
+	enumspb "go.temporal.io/api/enums/v1"
+	historypb "go.temporal.io/api/history/v1"
+	"go.temporal.io/api/workflowservice/v1"
+	"google.golang.org/grpc"
+)
+
+type serviceInterceptor struct {
+	encoders []PayloadEncoder
+}
+
+func (s *serviceInterceptor) processRequest(req interface{}) error {
+	switch r := req.(type) {
+	case *workflowservice.StartWorkflowExecutionRequest:
+		return encodePayloads(r.Input, s.encoders...)
+	case *workflowservice.SignalWorkflowExecutionRequest:
+		return encodePayloads(r.Input, s.encoders...)
+	case *workflowservice.SignalWithStartWorkflowExecutionRequest:
+		err := encodePayloads(r.Input, s.encoders...)
+		if err != nil {
+			return err
+		}
+
+		return encodePayloads(r.SignalInput, s.encoders...)
+	case *workflowservice.RespondActivityTaskCompletedRequest:
+		return encodePayloads(r.Result, s.encoders...)
+	case *workflowservice.RespondActivityTaskCompletedByIdRequest:
+		return encodePayloads(r.Result, s.encoders...)
+	case *workflowservice.RespondWorkflowTaskCompletedRequest:
+		return s.processCommands(r.Commands)
+	case *workflowservice.RecordActivityTaskHeartbeatRequest:
+		return encodePayloads(r.Details, s.encoders...)
+	case *workflowservice.RecordActivityTaskHeartbeatByIdRequest:
+		return encodePayloads(r.Details, s.encoders...)
+	}
+
+	return nil
+}
+
+func (s *serviceInterceptor) processResponse(response interface{}) error {
+	switch r := response.(type) {
+	case *workflowservice.GetWorkflowExecutionHistoryResponse:
+		return s.processEvents(r.History.Events)
+	case *workflowservice.PollWorkflowTaskQueueResponse:
+		if r.WorkflowType != nil {
+			return s.processEvents(r.History.Events)
+		}
+	case *workflowservice.PollActivityTaskQueueResponse:
+		if r.Input != nil {
+			err := decodePayloads(r.Input, s.encoders...)
+			if err != nil {
+				return err
+			}
+		}
+
+		if r.HeartbeatDetails != nil {
+			return decodePayloads(r.HeartbeatDetails, s.encoders...)
+		}
+	}
+
+	return nil
+}
+
+func (s *serviceInterceptor) processCommands(commands []*command.Command) error {
+	var err error
+	for _, c := range commands {
+		switch c.CommandType {
```

Comment:
A type switch is a bit safer. A command type set without the attributes will cause a panic here. Same for event type later.

Reply:
There are no types are there? afaik it's one type shared by all possible types of commands.

Reply:
There are, see the docs for https://pkg.go.dev/go.temporal.io/api/command/v1#Command. But I'm fine with what you have here, but may want to change `.Input` to `.GetInput()` which is safe from `GetXAttributes()` coming back nil.

Reply:
This is no longer relevant.